File: Online Judges/LeetCode/June LeetCoding Challenge/Week-4/Find the Duplicate Number.cpp
class Solution
{
public:
    int findDuplicate(vector<int> &nums)
    {
        int tortoise = nums[0], hare = nums[0];
        while (1)
        {
            tortoise = nums[tortoise];
            hare = nums[nums[hare]];
            if (hare == tortoise)
                break;
        }
        tortoise = nums[0];
        while (tortoise != hare)
        {
            tortoise = nums[tortoise];
            hare = nums[hare];
        }
        return hare;
    }
};

File: Online Judges/LeetCode/June LeetCoding Challenge/Week-4/Find the Duplicate Number.rs
impl Solution {
    pub fn find_duplicate(nums: Vec<i32>) -> i32 {
        let mut tortoise:i32 = nums[0];
        let mut hare:i32 = nums[0];
        
        loop {
            tortoise = nums[tortoise as usize];
            hare = nums[nums[hare as usize] as usize];
            if hare == tortoise {
                break;
            }
        }
        tortoise = nums[0];
        loop {
            if tortoise != hare {
                tortoise = nums[tortoise as usize];
                hare = nums[hare as usize];
            }
            else {
                return hare;
            }
        }       
    }
}